### PR TITLE
More suggestions for #8961

### PR DESCRIPTION
**Pull request by olix0r**
- Rename `port_names` to `tcp_ports_by_name` -- I know there was
  discussion about whether 'tcp' is ambiguous, but I think omitting it
  is more ambiguous. There's a difference between a *TCP port* and a
  *TCP probe*.
- Factor out a utility for reading a named port and verify that the port
  is actually TCP. This fixes a bug where we could reference a named UDP
  ports, for instance.
- Add several smaller tests to exercise some of these corner cases for
  issues that have been fixed.

Signed-off-by: Oliver Gould <ver@buoyant.io>

### CODE REVIEW

File: policy-controller/k8s/index/src/index.rs [hunk 1]
```diff
@@ -290,14 +290,19 @@ impl kubert::index::IndexNamespacedResource<k8s::Pod> for Index {
         let name = pod.name_unchecked();
         let _span = info_span!("apply", ns = %namespace, %name).entered();
 
-        let port_names = pod::port_names(&pod.spec);
-        let meta = pod::Meta::from_metadata(pod.metadata);
+        let port_names = pod
+            .spec
+            .as_ref()
+            .map(pod::tcp_ports_by_name)
+            .unwrap_or_default();
```
Comment (kleimkuhler) [inline policy-controller/k8s/index/src/index.rs:297]: 👍

**Review by kleimkuhler — APPROVED** [review]
Looks good! I like the intro of `fn named_tcp_port` and the extra test cases make sense.
